File: HB/20200828/큰수만들기.cpp
#include <string>
#include <vector>

using namespace std;

string solution(string number, int k)
{
    string answer = "";

    int n = number.size() - k;
    int start_idx = 0;
    for (int i = 0; i < n; i++)
    {
        char max_num = number[start_idx];
        int max_idx = start_idx;
        for (int j = start_idx; j <= k + i; j++)
        {
            if (max_num < number[j])
            {
                max_num = number[j];
                max_idx = j;
            }
        }
        start_idx = max_idx + 1;
        answer += max_num;
    }

    return answer;
}

File: JY/8/8.28/큰수만들기.py
from collections import deque
def solution(number, k):
    answer = ''
    n = len(number)
    if n==k:
        return '0'
    number = deque(number)
    stack = []
    cnt = 0
    stack.append(number[0])
    number.popleft()
    
    while cnt<=k and deque and number:
        now = number.popleft()
        while stack and cnt<k:
            if stack[-1]<now:
                stack.pop()
                cnt += 1
            else:
                break
        stack.append(now)
    if cnt!=k:
        stack = stack[:-(k-cnt)]
    answer = ''.join(stack)
            
    
    return answer